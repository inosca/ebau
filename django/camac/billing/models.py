from django.db import models


class BillingV2Entry(models.Model):
    TAX_MODE_INCLUSIVE = "inclusive"
    TAX_MODE_EXCLUSIVE = "exclusive"
    TAX_MODE_EXEMPT = "exempt"
    TAX_MODE_CHOICES = (
        (TAX_MODE_INCLUSIVE, "Incl 7.7%"),
        (TAX_MODE_EXCLUSIVE, "Excl 7.7%"),
        (TAX_MODE_EXEMPT, "Tax exempt"),
    )

    CALCULATION_FLAT = "flat"
    CALCULATION_PERCENTAGE = "percentage"
    CALCULATION_HOURLY = "hourly"
    CALCULATION_CHOICES = (
        (CALCULATION_FLAT, "Flat rate"),
        (CALCULATION_PERCENTAGE, "Percentage"),
        (CALCULATION_HOURLY, "Hourly"),
    )

    MUNICIPAL = "municipal"
    CANTONAL = "cantonal"
    ORGANIZATION_CHOICES = ((MUNICIPAL, "Municipal"), (CANTONAL, "Cantonal"))

    DECIMAL_FORMAT = {
        "max_digits": 10,
        "decimal_places": 2,
        "null": True,
        "blank": True,
    }

    # Structural: Which instance is the item billed to?
    instance = models.ForeignKey("instance.Instance", models.CASCADE, related_name="+")

    # Organisation: Who charged the item?
    group = models.ForeignKey("user.Group", models.DO_NOTHING, related_name="+")
    user = models.ForeignKey("user.User", models.DO_NOTHING, related_name="+")

    # Billing text
    text = models.TextField()
    date_added = models.DateField(auto_now_add=True)
    date_charged = models.DateField(null=True, blank=True)

    # Tax mode = calculation model for tax
    tax_mode = models.CharField(choices=TAX_MODE_CHOICES, max_length=20)

    # Calculation mode
    calculation = models.CharField(choices=CALCULATION_CHOICES, max_length=20)

    # Tax rate (percentage)
    tax_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Calculation mode: hourly rate
    hours = models.DecimalField(**DECIMAL_FORMAT)
    hourly_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Calculation mode: percentage of total cost
    percentage = models.DecimalField(**DECIMAL_FORMAT)
    total_cost = models.DecimalField(**DECIMAL_FORMAT)

    # Final rate (may be entered directly in "flat" mode, otherwise it's
    # calculated. We store it for easier handling in the output however.
    final_rate = models.DecimalField(**DECIMAL_FORMAT)

    # Organization: either municipal or cantonal but can be NULL
    # Used to distinguish which oranization collects part of the bill
    organization = models.CharField(
        choices=ORGANIZATION_CHOICES, max_length=20, null=True
    )
