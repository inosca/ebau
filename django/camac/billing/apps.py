from django.apps import AppConfig


class BillingConfig(AppConfig):
    name = "camac.billing"
