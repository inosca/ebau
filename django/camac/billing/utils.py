from collections import OrderedDict
from decimal import Decimal
from typing import List, TypedDict, Union

from camac.billing.models import BillingV2Entry


class OrganizationTotals(TypedDict):
    uncharged: str
    total: str


class BillingTotals(TypedDict):
    BillingV2Entry.CANTONAL: OrganizationTotals
    BillingV2Entry.MUNICIPAL: OrganizationTotals
    all: OrganizationTotals


def round_decimal(num: Decimal) -> Decimal:
    """Round decimal to 2 decimal places."""

    return num.quantize(Decimal("0.01"))


def calculate_final_rate(
    calculation: str,
    total_cost: Union[Decimal, None] = None,
    percentage: Union[Decimal, None] = None,
    hours: Union[Decimal, None] = None,
    hourly_rate: Union[Decimal, None] = None,
) -> Union[Decimal, None]:
    """Calculate final rate for given calculation type.

    - flat: use `total_cost`
    - percentage: `percentage` of `total_cost`
    - hourly: `hours` times `hourly_rate`
    """

    final_rate = None

    if calculation == BillingV2Entry.CALCULATION_FLAT:
        final_rate = total_cost
    elif calculation == BillingV2Entry.CALCULATION_PERCENTAGE:
        final_rate = total_cost * percentage / Decimal(100)
    elif calculation == BillingV2Entry.CALCULATION_HOURLY:
        final_rate = hours * hourly_rate

    # Don't ignore final_rate when value is 0
    return round_decimal(final_rate) if final_rate is not None else None


def add_taxes_to_final_rate(
    final_rate: Decimal, tax_mode: str, tax_rate: Decimal
) -> Union[Decimal, None]:
    """Add taxes to final rate.

    This only applies the `tax_rate` to the `final_rate` if the `tax_mode` is
    "exclusive".
    """

    # Don't ignore final_rate when value is 0
    if final_rate is None:
        return None

    if tax_mode != BillingV2Entry.TAX_MODE_EXCLUSIVE:
        return final_rate

    return round_decimal(final_rate + final_rate * tax_rate / Decimal(100))


def get_totals(entries: List[OrderedDict]) -> BillingTotals:
    """Get totals for a list of billing entries.

    This will return a dict of totals per organization type and over all
    organizations (including entries without an organization).
    """

    totals = {}

    for key, _ in BillingV2Entry.ORGANIZATION_CHOICES:
        totals[key] = get_totals_for_organization(entries, key)

    totals["all"] = get_totals_for_organization(entries)

    return totals


def get_totals_for_organization(
    entries: List[OrderedDict], organization: Union[str, None] = None
) -> OrganizationTotals:
    """Calculate total including uncharged total for an organization."""

    filtered_entries = [
        entry
        for entry in entries
        if not organization or entry["organization"] == organization
    ]

    return {
        "uncharged": str(
            round_decimal(
                Decimal(
                    sum(
                        [
                            Decimal(entry["final_rate"])
                            for entry in filtered_entries
                            if entry["date_charged"] is None
                            and entry["final_rate"]
                            is not None  # Don't ignore entry when final_rate is 0
                        ]
                    )
                )
            )
        ),
        "total": str(
            round_decimal(
                Decimal(
                    sum(
                        [
                            Decimal(entry["final_rate"])
                            for entry in filtered_entries
                            if entry["final_rate"]
                            is not None  # Don't ignore entry when final_rate is 0
                        ]
                    )
                )
            )
        ),
    }
