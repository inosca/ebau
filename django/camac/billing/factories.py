from decimal import Decimal

import pytz
from factory import Faker, LazyAttribute, Maybe, SubFactory, fuzzy
from factory.django import DjangoModelFactory

from camac.billing.models import BillingV2Entry
from camac.billing.utils import add_taxes_to_final_rate, calculate_final_rate
from camac.instance.factories import InstanceFactory
from camac.user.factories import GroupFactory, UserFactory


def choice_keys(choices: tuple):
    return [choice[0] for choice in choices]


class BillingV2EntryFactory(DjangoModelFactory):
    instance = SubFactory(InstanceFactory)
    user = SubFactory(UserFactory)
    group = SubFactory(GroupFactory)
    date_added = Faker("past_datetime", tzinfo=pytz.UTC)
    date_charged = None
    organization = fuzzy.FuzzyChoice(choice_keys(BillingV2Entry.ORGANIZATION_CHOICES))
    billing_type = fuzzy.FuzzyChoice(choice_keys(BillingV2Entry.BILLING_TYPE_CHOICES))
    text = Faker("word")
    legal_basis = Faker("word")
    cost_center = Faker("aba")

    tax_mode = fuzzy.FuzzyChoice(choice_keys(BillingV2Entry.TAX_MODE_CHOICES))
    tax_rate = Maybe(
        "is_tax_exempt",
        yes_declaration=Decimal(0),
        no_declaration=fuzzy.FuzzyChoice([Decimal(2.5), Decimal(7.7)]),
    )

    calculation = fuzzy.FuzzyChoice(choice_keys(BillingV2Entry.CALCULATION_CHOICES))
    total_cost = Maybe(
        "is_flat_or_percentage",
        yes_declaration=Faker(
            "pydecimal", left_digits=3, right_digits=2, positive=True
        ),
        no_declaration=None,
    )
    percentage = Maybe(
        "is_percentage",
        yes_declaration=Faker("pydecimal", min_value=1, max_value=100),
        no_declaration=None,
    )
    hours = Maybe(
        "is_hourly",
        yes_declaration=Faker(
            "pydecimal", left_digits=1, right_digits=0, positive=True
        ),
        no_declaration=None,
    )
    hourly_rate = Maybe(
        "is_hourly",
        yes_declaration=Faker(
            "pydecimal", left_digits=3, right_digits=0, positive=True
        ),
        no_declaration=None,
    )

    final_rate = LazyAttribute(
        lambda e: add_taxes_to_final_rate(
            calculate_final_rate(
                calculation=e.calculation,
                total_cost=e.total_cost,
                percentage=e.percentage,
                hours=e.hours,
                hourly_rate=e.hourly_rate,
            ),
            tax_mode=e.tax_mode,
            tax_rate=e.tax_rate,
        )
    )

    class Meta:
        model = BillingV2Entry

    class Params:
        is_flat = LazyAttribute(
            lambda e: e.calculation == BillingV2Entry.CALCULATION_FLAT
        )
        is_percentage = LazyAttribute(
            lambda e: e.calculation == BillingV2Entry.CALCULATION_PERCENTAGE
        )
        is_hourly = LazyAttribute(
            lambda e: e.calculation == BillingV2Entry.CALCULATION_HOURLY
        )
        is_flat_or_percentage = LazyAttribute(lambda e: e.is_flat or e.is_percentage)
        is_tax_exempt = LazyAttribute(
            lambda e: e.tax_mode == BillingV2Entry.TAX_MODE_EXEMPT
        )
