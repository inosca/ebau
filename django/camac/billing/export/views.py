import django_excel as excel
from rest_framework.exceptions import ErrorDetail
from rest_framework.generics import ListAPIView
from rest_framework.renderers import BaseRenderer
from rest_framework.settings import api_settings
from rest_framework_json_api import renderers as jsonapi_renderers

from camac.billing.filters import (
    BillingV2EntryExportFilterBackend,
    BillingV2EntryFilterSet,
)
from camac.billing.models import BillingV2Entry
from camac.billing.serializers import BillingV2EntryExportSerializer
from camac.instance.mixins import InstanceQuerysetMixin


class XLSXRenderer(jsonapi_renderers.JSONRenderer, BaseRenderer):
    """
    Excel output renderer for the REST framework.

    This formats the data into an Excel table (instead of JSON for example) as part
    of the regular REST framework infrastructure.

    Use this by setting `renderer_classes = [XLSXRenderer]` in your view class.
    """

    media_type = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
    format = "xlsx"

    def render(self, data, accepted_media_type=None, renderer_context=None):
        view = renderer_context["view"]
        serializer = view.get_serializer()
        fields = serializer.get_fields()
        columns = list(fields)

        is_error = False
        try:
            is_error = (
                isinstance(data[0]["detail"], ErrorDetail) if len(data) else False
            )
        except KeyError:
            pass

        if is_error:
            return super().render(data)

        header = [field.label or name for name, field in fields.items()]

        rows = [[row[col] for col in columns] for row in data]

        sheet = excel.pe.Sheet([header] + rows)

        return excel.make_response(sheet, file_type="xlsx", file_name="list.xlsx")


class BillingV2EntryExportView(InstanceQuerysetMixin, ListAPIView):
    queryset = BillingV2Entry.objects
    serializer_class = BillingV2EntryExportSerializer
    filter_backends = api_settings.DEFAULT_FILTER_BACKENDS + [
        BillingV2EntryExportFilterBackend,
    ]
    filterset_class = BillingV2EntryFilterSet
    renderer_classes = [XLSXRenderer]

    # Queryset for internal role permissions are handled
    # by InstanceQuerysetMixin
    def get_queryset_for_applicant(self):
        return self.queryset.none()

    def get_queryset_for_public(self):
        return self.queryset.none()
