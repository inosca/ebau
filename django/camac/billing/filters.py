from caluma.caluma_form.models import Answer
from django.conf import settings
from django.db.models import (
    CharField,
    OuterRef,
    Value,
)
from django.db.models.functions import Cast, Coalesce, NullIf, Replace, Trim
from django.utils.translation import gettext
from django_filters.rest_framework import DateFilter, FilterSet
from rest_framework.filters import BaseFilterBackend
from rest_framework.serializers import ValidationError

from camac.billing.models import BillingV2Entry
from camac.instance.export.filters import ConcatWS, StringAggSubquery


class BillingV2EntryFilterSet(FilterSet):
    date_added_before = DateFilter(field_name="date_added", lookup_expr="lte")
    date_added_after = DateFilter(field_name="date_added", lookup_expr="gte")

    class Meta:
        model = BillingV2Entry
        fields = [
            "instance",
            "date_added_before",
            "date_added_after",
        ]


class BillingV2EntryExportFilterBackend(BaseFilterBackend):
    def _validate_qs_size(self, queryset):
        if queryset.count() > settings.BILLING_EXPORT_MAX_RECORDS:  # pragma: no cover
            raise ValidationError(
                gettext(
                    "Too many records in output. Reduce "
                    "data range using filters to below {max}",
                ).format(
                    max=settings.BILLING_EXPORT_MAX_RECORDS,
                )
            )

    def filter_queryset(self, request, queryset, view):
        self._validate_qs_size(queryset)

        def get_answer(
            slug: str,
            ref: str = "instance__case__document_id",
            all_answers: bool = False,
        ):
            """
            Fetch answers filtered by a question ID and document reference, optionally fetching all answers.

            Args:
                slug (str): The question ID to filter answers.
                ref (str, optional): The reference field for the document. Defaults to 'instance__case__document_id'.
                all_answers (bool, optional): If True, fetch all answers by document family,
                                            otherwise fetch the first one by document ID. Defaults to False.

            Returns:
                QuerySet: A queryset containing the string values of the answers.
            """
            document_key = "document__family" if all_answers else "document_id"
            filter_kwargs = {
                "question_id": slug,
                document_key: OuterRef(ref),
            }

            queryset = (
                Answer.objects.filter(**filter_kwargs)
                .annotate(
                    string_value=NullIf(
                        Trim(
                            Replace(
                                Cast("value", output_field=CharField()),
                                Value('"'),
                                Value(""),
                            )
                        ),
                        Value(""),
                    )
                )
                .values("string_value")
            )

            # We don't join to the answerdoc, but enforce explicit order by
            # rowdocument's creation date. This is precise enough to ensure
            # all columns in output will be sorted in the same way
            queryset = queryset.order_by("document__created_at")

            return queryset if all_answers else queryset[:1]

        # We need to put a `NullIf` function around the street and city in order
        # to filter them out properly if empty. This is needed because
        # `CONCAT_WS` always returns a string, even if all concatenated values
        # are empty.
        address = Coalesce(
            get_answer("standort-migriert"),
            ConcatWS(
                NullIf(
                    ConcatWS(
                        get_answer("strasse-flurname"),
                        get_answer("nr"),
                        delimiter=" ",
                    ),
                    Value(""),
                ),
                NullIf(
                    ConcatWS(
                        get_answer("plz-grundstueck-v3"),
                        get_answer("ort-grundstueck"),
                        delimiter=" ",
                    ),
                    Value(""),
                ),
                delimiter=", ",
            ),
        )

        parcels = StringAggSubquery(
            get_answer(slug="parzellennummer", all_answers=True),
            column_name="string_value",
            delimiter=", ",
        )

        coordinate_X = StringAggSubquery(
            get_answer(slug="lagekoordinaten-ost", all_answers=True),
            column_name="string_value",
            delimiter=", ",
        )

        coordinate_Y = StringAggSubquery(
            get_answer(slug="lagekoordinaten-nord", all_answers=True),
            column_name="string_value",
            delimiter=", ",
        )

        return queryset.annotate(
            address=address,
            parcels=parcels,
            coordinate_X=coordinate_X,
            coordinate_Y=coordinate_Y,
        )
