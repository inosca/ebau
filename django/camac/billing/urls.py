from django.urls import re_path
from rest_framework.routers import SimpleRouter

from camac.billing import views
from camac.billing.export.views import BillingV2EntryExportView

urlpatterns = [
    re_path(
        r"billing-v2-entries/export$",
        BillingV2EntryExportView.as_view(),
        name="billing-export",
    ),
]

r = SimpleRouter(trailing_slash=False)

r.register(r"billing-v2-entries", views.BillingV2EntryViewset, "billing-v2-entry")

urlpatterns.extend(r.urls)
