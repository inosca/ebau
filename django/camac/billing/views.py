from django.db import transaction
from django.utils import timezone
from rest_framework import response, status
from rest_framework.decorators import action
from rest_framework_json_api.views import ModelViewSet

from camac.billing.filters import BillingV2EntryFilterSet
from camac.billing.models import BillingV2Entry
from camac.billing.serializers import BillingV2EntrySerializer
from camac.instance.mixins import InstanceQuerysetMixin
from camac.user.permissions import permission_aware


class BillingV2EntryViewset(ModelViewSet, InstanceQuerysetMixin):
    swagger_schema = None
    serializer_class = BillingV2EntrySerializer
    filterset_class = BillingV2EntryFilterSet
    queryset = BillingV2Entry.objects.all().order_by("organization", "pk")

    @permission_aware
    def get_queryset(self):
        return self.queryset.none()

    def get_queryset_for_municipality(self):
        return super().get_queryset()

    def get_queryset_for_service(self):
        return super().get_queryset()

    def has_object_charge_permission(self, obj):
        return (
            obj.instance.responsible_service(filter_type="municipality")
            == self.request.group.service
        )

    @action(methods=["PATCH"], detail=True)
    @transaction.atomic
    def charge(self, request, pk=None):
        billing_entry = self.get_object()
        billing_entry.date_charged = timezone.now().date()
        billing_entry.save(update_fields=["date_charged"])

        return response.Response(status=status.HTTP_204_NO_CONTENT)
