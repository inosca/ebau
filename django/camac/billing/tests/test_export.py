from functools import partial

import pyexcel
import pytest
from dateutil.relativedelta import relativedelta
from django.urls import reverse
from django.utils import timezone
from rest_framework import status

from camac.billing.export.views import BillingV2EntryExportView
from camac.billing.models import BillingV2Entry


@pytest.mark.freeze_time("2023-11-06")
@pytest.mark.parametrize(
    "parcels",
    [
        # List of number, X, Y coord
        [  # No parcels in dossier
        ],
        [
            # Single-parcel dossier
            (134, 2222, 3333)
        ],
        [
            # Multi-parcel dossier
            (231, 1111, 2222),
            (232, 4444, 5555),
            (233, 6666, 7777),
        ],
    ],
)
@pytest.mark.parametrize(
    "role__name,billing_params,entries_count, expected_count, filter_date_added",
    [
        ("Municipality", {"calculation": BillingV2Entry.CALCULATION_FLAT}, 3, 3, False),
        (
            "Municipality",
            {"calculation": BillingV2Entry.CALCULATION_PERCENTAGE},
            3,
            3,
            False,
        ),
        (
            "Municipality",
            {
                "calculation": BillingV2Entry.CALCULATION_HOURLY,
                "tax_mode": BillingV2Entry.TAX_MODE_INCLUSIVE,
            },
            3,
            3,
            False,
        ),
        (
            "Municipality",
            {
                "calculation": BillingV2Entry.CALCULATION_FLAT,
                "tax_mode": BillingV2Entry.TAX_MODE_EXEMPT,
            },
            3,
            3,
            False,
        ),
        ("Applicant", {"calculation": BillingV2Entry.CALCULATION_FLAT}, 3, 0, False),
        ("Support", {"calculation": BillingV2Entry.CALCULATION_FLAT}, 3, 3, False),
        ("Service", {"calculation": BillingV2Entry.CALCULATION_FLAT}, 3, 3, False),
        ("Service", {"calculation": BillingV2Entry.CALCULATION_FLAT}, 3, 1, True),
    ],
)
def test_billing_export(
    admin_client,
    billing_v2_entry_factory,
    snapshot,
    instance_with_document_for_billing,
    billing_params,
    entries_count,
    expected_count,
    filter_date_added,
    parcels,
):
    instance, parcel_ans = instance_with_document_for_billing

    billing_v2_entry_factory.create_batch(
        entries_count, **{**billing_params, "instance": instance}
    )
    query_params = {"instance": instance.pk}

    for idx, (parcel_nr, coord_x, coord_y) in enumerate(parcels):
        row_doc = parcel_ans.documents.create(
            through_defaults={"sort": idx},
            form=parcel_ans.question.row_form,
            family_id=parcel_ans.document_id,
        )
        row_doc.answers.create(question_id="parzellennummer", value=parcel_nr)
        row_doc.answers.create(question_id="lagekoordinaten-ost", value=coord_x)
        row_doc.answers.create(question_id="lagekoordinaten-nord", value=coord_y)

    if filter_date_added:
        date_added = timezone.now()  # freezed
        date_later = (date_added + relativedelta(months=1)).date()
        entry = BillingV2Entry.objects.first()
        entry.date_added = date_later
        entry.save()
        query_params = {**query_params, "filter[date_added_after]": date_later}

    url = reverse("billing-export")
    response = admin_client.get(url, query_params)

    assert response.status_code == status.HTTP_200_OK
    book = pyexcel.get_book(file_content=response.content, file_type="xlsx")
    sheet = book.get_dict()["pyexcel sheet"]
    assert len(sheet) - 1 == expected_count  # substract header row
    if len(sheet) > 1:
        data = sheet[1]
        data.pop(7)  # remove ebau-Nr as it is not stable

        snapshot.assert_match(data)


@pytest.fixture
def instance_with_document_for_billing(
    instance,
    question_factory,
    answer_factory,
    form_factory,
    case_factory,
    form_question_factory,
):
    instance.case = case_factory()
    instance.save()
    parcel_table = question_factory(type="table")
    parcel_field = partial(
        form_question_factory, form=parcel_table.row_form, question__type="integer"
    )
    parcel_field(question__slug="parzellennummer")
    parcel_field(question__slug="lagekoordinaten-ost")
    parcel_field(question__slug="lagekoordinaten-nord")
    instance.case.document.form.questions.add(parcel_table)

    parcel_ans = answer_factory(document=instance.case.document, question=parcel_table)
    return instance, parcel_ans


@pytest.mark.parametrize(
    "role__name,method,has_access,expected_count",
    [
        ("Municipality", "get_queryset_for_municipality", True, 1),
        ("Service", "get_queryset_for_service", True, 1),
        ("Applicant", "_get_queryset_for_applicant", True, 0),
        ("Public", "get_queryset_for_public", True, 0),
        ("Municipality", "get_queryset_for_municipality", False, 0),
        ("Service", "get_queryset_for_service", False, 0),
        ("Applicant", "_get_queryset_for_applicant", False, 0),
        ("Public", "get_queryset_for_public", False, 0),
    ],
)
def test_billing_entry_export_visibilities(
    db,
    admin_client,
    instance,
    mocker,
    billing_v2_entry_factory,
    role,
    group,
    method,
    expected_count,
    has_access,
):
    is_public = role.name == "Public"
    mocker.patch(
        "camac.user.permissions.get_group", return_value=None if is_public else group
    )
    billingv2entry = billing_v2_entry_factory(instance=instance)
    mocker.patch(
        f"camac.instance.mixins.InstanceQuerysetMixin.{method}",
        return_value=BillingV2Entry.objects.filter(pk=billingv2entry.pk)
        if has_access
        else BillingV2Entry.objects.none(),
    )

    view = BillingV2EntryExportView()
    assert view.get_queryset().count() == expected_count
    if expected_count:
        assert billingv2entry in view.get_queryset()
