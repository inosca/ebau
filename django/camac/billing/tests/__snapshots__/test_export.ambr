# serializer version: 1
# name: test_billing_export[Municipality-billing_params0-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '297.28 exklusive  7.7% MwSt',
    '297.28',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Municipality-billing_params1-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '29.86 von 276.03 exklusive  7.7% MwSt',
    '88.79',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Municipality-billing_params2-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '3.00 Std à 828.00 inklusive  7.7% MwSt',
    '2484.00',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Municipality-billing_params3-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '276.03 nicht MwSt-pflichtig',
    '276.03',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Service-billing_params6-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '297.28 exklusive  7.7% MwSt',
    '297.28',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Service-billing_params7-3-1-True]
  list([
    '06.12.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '297.28 exklusive  7.7% MwSt',
    '297.28',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
# name: test_billing_export[Support-billing_params5-3-3-False]
  list([
    '06.11.2023',
    'audience',
    'Jeffrey Zhang',
    'Rebecca Gonzalez',
    '297.28 exklusive  7.7% MwSt',
    '297.28',
    '',
    '',
    '',
    'Keine Koordinaten verfügbar',
    'Alex Smith',
  ])
# ---
